src/MongoDB.Driver/Linq/Translators/PredicateTranslator.cs

```diff
@@ -217,6 +217,32 @@ private bool CanAnyBeRenderedWithoutElemMatch(Expression node)
             }
         }
 
+        private FilterDefinition<BsonDocument> ConvertElemMatchFilterToScalarElementMatchIfNeeded(FilterDefinition<BsonDocument> filter, IFieldExpression fieldExpression, Expression wherePredicate)
+        {
+            if ((!(fieldExpression.Serializer is IBsonDocumentSerializer)) || DoesExpressionUseDocumentItself(wherePredicate))
+            {
+                return new ScalarElementMatchFilterDefinition<BsonDocument>(filter);
+            }
+            else
+            {
+                return filter;
+            }
+        }
+
+        private bool DoesExpressionUseDocumentItself(Expression node)
+        {
+            // if a left operand is DocumentExpression, we need to generate a "$elemMatch" in a short form,
+            // otherwise we will have $elemMatch queries with gaps similar to : "{ $elemMatch : { ' ' : {"
+            if (node is BinaryExpression binaryExpression)
```

Comment:
How do we know for sure that only BinaryExpression needs special consideration?

Reply:
I can't imagine any checking to be 100% sure. But
1. In case if not the `BinaryExpression`, we will have `legacy` behavior. So, none new bugs will be added. The same time as the customer problem will be fixed.
2. As an example of not `BinaryExpression`, it can be this query:

>x => x.G.Any(g => true)

the generated query will be like this:

>{G={ "$elemMatch" : { } }}

the expression type will be `ConstantExpression`. It's the legacy "bug". Not sure, I guess the expected solution here, do not use constants in the `predicate`, since it can be just:

>x => x.G.Any()

Currently, I cannot imagine any other query which will give `true`/`false` results which are required from a predicate in LINQ `Any`. @rstam, what do you think?

Reply:
I don't know... but if you are comfortable that BinaryExpression is the only thing we need to handle specially I'll take your word for it.